### DISCUSSION TITLE
Folding at keywords and whitespace

### DISCUSSION THREAD

**Discussion by neilsf** [orbitalquark/textadept#117] [Q&A] [answered]
Hi
Disclaimer: I am not a Lua programmer.
I'm trying to write a custom lexer for a BASIC dialect that allows code blocks like this:

    SUB something
        ...
    END SUB

Now if I define the keywords "SUB" and "END" then syntax highlighting works fine, however I can't find a way to define folding.
I tried this:

    lex:add_fold_point(lexer.KEYWORD, 'sub', 'end sub')

It doesn't work, I think because "end sub" is not a keyword. I also tried:

    lex:add_fold_point(lexer.KEYWORD, 'sub', 'end')

It doesn't work either, I think becuase there is text content after "end".

I also tried to add keywords containing a space like this:

    lex:add_rule('keyword', token(lexer.KEYWORD, lexer.word_match({'sub', 'end sub'}, true)))

This also doesn't work, it gives an error message when launching Textadept.

    /home/neils/Programs/textadept_10.8.x86_64/lexers/lexer.lua:1760: bad argument #1 to 'lpeg_S' (string expected, got boolean)

Can you give me ideas on how it could be written? Thanks

**Comment by orbitalquark** [comment]
Did you try

lex:add_fold_point(lexer.KEYWORD, 'SUB', 'END')

Lexers are case-sensitive. If your keywords are for 'SUB' and 'END', then the fold point should use the same case.

The error you are getting in that last example is due to the fact that that API is valid for 11.x, but not 10.8.

**Reply by neilsf** [reply]
> Did you try lex:add_fold_point(lexer.KEYWORD, 'SUB', 'END')

I did but it only works if `END` is the only keyword in that line. It doesn't work for `END SUB` or `END FUNCTION` etc...

> The error you are getting in that last example is due to the fact that that API is valid for 11.x, but not 10.8.

Got that, I have upgraded to the latest version, thanks.

**Comment by orbitalquark** [comment, marked as answer]
Oh, I see. 'End Sub' is broken up into two parts: 'End', which is a fold end point, but the very next 'Sub' starts a new fold point. Other than writing your own fold function, this is not possible right now. I'll have to look into this.

**Reply by neilsf** [reply]
Oh yeah, I see, great explanation, thanks.
I'll try to write a custom fold function, it's time to learn Lua anyway :-)

**Comment by orbitalquark** [comment]
This should be fixed via https://github.com/orbitalquark/scintillua/commit/83b52b0105b6a383efdeeb4488cbbcf902bc5ad6 in the Scintillua project (Textadept uses it for syntax highlighting). These changes should be in the next nightly build of Textadept.

**Reply by neilsf** [reply]
It works nicely in latest nightly, many thanks.
